### PR TITLE
Introduce local JWT validation, change credential cache key to podUID

### PR DESCRIPTION
**Pull request by gargipanatula**
*Issue #, if available:*

The PIA caches pod credentials keyed by JWT token. A pod’s JWT token is rotated every ~19 hours. When the kubelet rotates a pod's service account token, the PIA treats the new token as a cache miss and calls the Auth Service to validate the new token and get a fresh set of credentials — even when cached credentials on the old token are still valid. If the Auth Service is unavailable during rotation, the pod loses credential access despite there being valid credentials in the cache. This breaks static stability.

*Description of changes:*

The goal of local JWT validation is to prevent the PIA from calling the Auth Service each time a token rotates. Instead, when a pod has valid credentials in the cache, the PIA will attempt to validate the token locally. Validation occurs in two parts. First, the PIA will fetch the apiserver’s public signing keys (through the apiserver's /openid/v1/jwks endpoint) to cryptographically validate the token. Then, it will check claims of the token. The first time a pod makes a request to the PIA, its token will still be validated by the Auth Service since the Auth Service checks some claims that the PIA does not. However, future rotated tokens from a pod will be validated by the agent. In case the agent cannot validate a token, it will fall back to the Auth Service as a fail-safe.

The feature has two coordinated changes:
1. Pod UID–based cache key — credential cache keyed by pod UID (stable for pod lifetime) instead of the JWT string
2. Local JWT validation — claims check + cryptographic signature verification against JWKS public keys

The feature is fail-open - any signature or claims check failures will fall back to the Auth Service (barring expired tokens). It is also feature gated to 1.34+ since older clusters do not return the full set of keys in the `/openid/v1/jwks` endpoint ([KEP-740](https://github.com/kubernetes/enhancements/issues/740)).

Major callouts:
* The PIA's SA token is now automounted to the pod
* The PIA now writes to disk via emptyDir (to store the public keys)
* The PIA now calls the apiserver 

*Testing*
- E2E tests that mock valid and invalid JWTs and see whether a) local JWT validation occurred and b) if local validation had the expected result [pass/fail].
- Unit tests that ensure pod UID caching works — cache lookups use pod UID as the key, cache hits/misses behave correctly across token rotations for the same pod, and pod UID extraction from JWT claims handles malformed tokens gracefully.
- Unit tests that ensure all claims are parsed and checked correctly — validates exp, nbf, aud, and kubernetes.io identity claims (namespace, service account, pod name/UID), including edge cases like nil maps, missing nested fields, and wrong types.
- Unit tests that ensure cryptographic signature verification against JWKS public keys accepts validly-signed tokens and rejects tokens signed with unknown or incorrect keys.
- Unit tests that ensure JWKS disk persistence writes keys atomically on fetch and reads them back correctly on startup when the apiserver is unreachable.
- Unit tests that ensure version gating blocks local validation and JWKS disk fallback on clusters below K8s 1.34.
- Unit tests that ensure fail-open behavior — validation failures (excluding expired tokens) fall back to the Auth Service rather than rejecting the request.

This branch includes many changes, here's a breakdown of the commits and their reviewers:

- **[#114](https://github.com/aws/eks-pod-identity-agent/pull/114) — Change internal cache key to pod UID**
  - **Description:** Re-keys the PIA's credential cache from JWT string → pod UID. Since pod UID is constant for a pod's lifetime, cached credentials remain accessible even when JWTs rotate. Foundational change enabling static stability.
  - **Commits:**
    - [`7a63f9e`](https://github.com/aws/eks-pod-identity-agent/commit/7a63f9e7182636622ba4f3180e49263a8d301e11) — Change internal cache key to pod UID
    - [`7eb193a`](https://github.com/aws/eks-pod-identity-agent/commit/7eb193aafa19ac0945d0f3ce08507510bf920480) — Add comments to AuthServiceFailure test documenting flow
  - **Approved by:** @kmala
  - **Merged:** Jan 16, 2026

- **[#135](https://github.com/aws/eks-pod-identity-agent/pull/135) — Add internal JWT validation**
  - **Description:** Adds local JWT validation — fetches apiserver's JWKS public signing keys from `/openid/v1/jwks`, validates rotated tokens by checking claims (exp, nbf, k8s identity claims) and verifying cryptographic signatures. Falls back to Auth Service on failure. Only enabled on K8s ≥ 1.34.
  - **Commits:**
    - [`685dd5b`](https://github.com/aws/eks-pod-identity-agent/commit/685dd5bafcda52684e7cd5b44495910f8556b062) — internal jwt validation
    - [`43a937b`](https://github.com/aws/eks-pod-identity-agent/commit/43a937bd1d5d2abf90ae3b62b745245ff62133d9) — Add tests for edge cases
  - **Approved by:** @SaranBalaji90 (/lgtm + /approve)
  - **Reviewed/commented by:** @yue9944882, @a-cool-train, @SaranBalaji90
  - **Merged:** May 12, 2026

- **[#150](https://github.com/aws/eks-pod-identity-agent/pull/150) — Add JWK disk persistence**
  - **Description:** Persists JWKS public keys to disk (`emptyDir`) after every successful fetch so the agent can validate tokens after container restart when the apiserver is unreachable. Atomic writes with 0600 permissions. Graceful degradation if disk is unavailable.
  - **Commits:**
    - [`07b017b`](https://github.com/aws/eks-pod-identity-agent/commit/07b017be135467fd51a75ae140d2fe7ed0ec43a8) — add JWKS keys disk persistence
  - **Approved by:** @a-cool-train
  - **Merged:** May 14, 2026

- **[#151](https://github.com/aws/eks-pod-identity-agent/pull/151) — Add audience validation for service account tokens**
  - **Description:** Validates the token's `aud` claim contains `pods.eks.amazonaws.com` during local JWT validation to prevent confused-deputy attacks. Skipped when `--endpoint` is overridden (custom delegates may use different audiences).
  - **Commits:**
    - [`de9c987`](https://github.com/aws/eks-pod-identity-agent/commit/de9c987fde6a69f83349312a87f6166c5c618105) — Add audience validation for service account tokens
  - **Approved by:** @a-cool-train
  - **Merged:** May 15, 2026

- **[#152](https://github.com/aws/eks-pod-identity-agent/pull/152) — Block JWKS disk fallback and local validation on unsupported K8s versions**
  - **Description:** Gates local validation and JWKS disk fallback behind a K8s version check (≥ 1.34). Calls `/version` before triggering local validation. Adds panic recovery in `ValidateToken`. Prevents using stale/static keys on older clusters where `/openid/v1/jwks` doesn't reflect rotating external signers (pre-KEP-740).
  - **Commits:**
    - [`ffaf7c3`](https://github.com/aws/eks-pod-identity-agent/commit/ffaf7c34e352bdd2e86aaad367c286de1b2a4075) — block jwks disk fallback on unsupported versions
    - [`1232927`](https://github.com/aws/eks-pod-identity-agent/commit/1232927de6be4b517a303a736bf8f9b28f21e6be) — check version before triggering local validation
  - **Approved by:** @a-cool-train
  - **Reviewed/commented by:** @a-cool-train
  - **Merged:** Jun 9, 2026

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: internal/validation/token_validator.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	}
+	return nil
+}
+
+// ValidateToken performs claims validation and cryptographic signature
+// verification on the request's service account token.
+func (tv *TokenValidator) ValidateToken(ctx context.Context, req *credentials.EksCredentialsRequest) (err error) {
+	log := logger.FromContext(ctx)
+
+	defer func() {
```
Comment (SaranBalaji90) [inline internal/validation/token_validator.go:75]: Adding panic handler inside the helper methods sets the expectation that anyone adding a helper or any function should handle the panic inside their methods explicitly. If we really needed one we should just have one at the top level, not in every handler

File: internal/validation/jwk_store.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	tmp := f.Name()
+
+	// Write data to temp file; close + remove on failure to avoid fd/file leaks.
+	if _, err := f.Write(data); err != nil {
+		f.Close()
+		os.Remove(tmp)
+		return fmt.Errorf("failed to write temp file: %w", err)
+	}
+
+	// Close before rename so data is fully flushed to disk.
```
Comment (SaranBalaji90) [inline internal/validation/jwk_store.go]: Missing f.sync()?

File: internal/validation/jwk_store.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to marshal JWKSet: %w", err)
+	}
+
+	// Create temp file in the same directory
+	f, err := os.CreateTemp(dir, ".jwks-cache-*.tmp")
+	if err != nil {
+		return fmt.Errorf("failed to create temp file: %w", err)
+	}
+	defer f.Close() // safety net in case a panic skips the explicit closes below
```
Comment (SaranBalaji90) [inline internal/validation/jwk_store.go:33]: You are explicitly close in line 44?

File: internal/credsretriever/refreshing_cache.go [hunk 4]
```diff
@@ -191,10 +201,100 @@ func (r *cachedCredentialRetriever) GetIamCredentials(ctx context.Context,
...
+	// If the cached credentials' token matches the incoming requests' token, return the credentials
+	if val.originatingRequest.ServiceAccountToken == request.ServiceAccountToken {
+		log.WithField("cache-hit", 1).Tracef("Using cached credentials")
+		return val.credentials, true
+	}
+
+	// Otherwise, attempt to validate the token locally
+	tv, ok := r.tokenValidator.Load().(tokenValidator)
+	if !ok {
+		r.tryInitTokenValidator(ctx)
```
Comment (SaranBalaji90) [inline internal/credsretriever/refreshing_cache.go]: Right way to do this is to create a separate ctx rather than using the one from pod reqeust path

File: internal/credsretriever/refreshing_cache.go [hunk 5]
```diff
@@ -260,7 +360,12 @@ func (r *cachedCredentialRetriever) onCredentialRenewal(key string, entry cacheE
 		if isIrrecoverableError {
 			log.Infof("Removing credentials from cache, got non recoverable error: %s", err.Error())
 			promCacheError.WithLabelValues("NonRecoverable", errCode).Inc()
-			r.internalCache.Delete(entry.originatingRequest.ServiceAccountToken)
+			podUID, err := getPodUIDfromServiceAccountToken(entry.originatingRequest.ServiceAccountToken)
```
Comment (SaranBalaji90) [inline internal/credsretriever/refreshing_cache.go:363]: Why cant we use the key directly and need to recompute?

File: internal/credsretriever/refreshing_cache.go [hunk 6]
```diff
@@ -191,10 +201,100 @@ func (r *cachedCredentialRetriever) GetIamCredentials(ctx context.Context,
...
+		time.Sleep(defaultActiveRequestWaitTime)
+	}
+	return true
+}
+
 func (r *cachedCredentialRetriever) callDelegateAndCache(ctx context.Context,
 	request *credentials.EksCredentialsRequest) (cacheEntry, credentials.ResponseMetadata, error) {
 	log := logger.FromContext(ctx)
 
+	podUID, err := getPodUIDfromServiceAccountToken(request.ServiceAccountToken)
```
Comment (SaranBalaji90) [inline internal/credsretriever/refreshing_cache.go:293]: looks like caller already has the podUUID

File: internal/validation/token_validator.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+	log := logger.FromContext(ctx)
+	newKeys := make(keyCache)
+	for _, k := range jwks.Keys {
+		if _, dup := newKeys[k.Kid]; dup {
+			continue
+		}
+		pubKey, err := parseJWK(k)
+		if err != nil {
+			log.Infof("skipping key %s: %v", k.Kid, err)
+			continue
```
Comment (SaranBalaji90) [inline internal/validation/token_validator.go:143]: Make sure we have test coverage wherever needed

File: internal/validation/token_validator.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+		}
+	}()
+
+	if req.ServiceAccountToken == "" {
+		return pkgerrors.NewRequestValidationError("Service account token cannot be empty")
+	}
+
+	if tv.jwksSource == nil {
+		return fmt.Errorf("jwksSource is nil, skipping local validation")
+	}
```
Comment (SaranBalaji90) [inline internal/validation/token_validator.go:88]: Given these are standalone helper functions, it might be good to have test coverage to ensure automation doesnt remove these checks and causes nil pointer at some other place

File: internal/validation/jwk_store.go (hunk 2 above)
Comment (gargipanatula) [reply internal/validation/jwk_store.go]: I think the comment is misleading so will update it in next commit.

The intent here is to protect the data in case of the container/process crashing, which f.Write + f.Close already guarantees (data is in the kernel page cache). We don't need to force a write to disk w/ `f.Sync`.

File: internal/validation/token_validator.go (hunk 8 above)
Comment (gargipanatula) [reply internal/validation/token_validator.go:88]: Done - added 4 tests in `TestValidateToken` in `internal/validation/token_validator_test.go`. 

1. empty token returns error
2. nil jwksSource returns error — constructs a TokenValidator with no jwksSource set
3. unparseable token returns error
4. k8s version check failure returns error

File: internal/validation/token_validator.go (hunk 7 above)
Comment (gargipanatula) [reply internal/validation/token_validator.go:143]: Done - added tests for `loadJWKSet` in `TestLoadJWKSet` in `token_validator_test.go`. 
  
  1. valid key is loaded — happy path
  2. empty JWKSet results in empty cache — no keys, no panic
  3. unparseable key is skipped — unsupported key type gets skipped gracefully
  4. unparseable key skipped but valid key still loaded — bad keys don't poison good ones
  5. duplicate kid keeps first key only — dedup logic works
  6. all keys fail to parse results in empty cache — bad base64 + bad curve, cache is empty but no crash

File: internal/credsretriever/refreshing_cache.go (hunk 4 above)
Comment (gargipanatula) [reply internal/credsretriever/refreshing_cache.go]: Will fix in next commit

File: internal/validation/token_validator.go (hunk 1 above)
Comment (gargipanatula) [reply internal/validation/token_validator.go:75]: Discussed offline - agreed that it may be better to move to top level, but given it adds complexity to the existing path, this may be better as a post-merge enhancement given the existing size of this change.

File: internal/credsretriever/refreshing_cache.go (hunk 5 above)
Comment (gargipanatula) [reply internal/credsretriever/refreshing_cache.go:363]: Discussed offline - follows pattern of existing code (which also parsed the request's token). Kept as is to avoid adding change/complexity to the PR but will consider fixing as a post-merge enhancement.

File: internal/credsretriever/refreshing_cache.go (hunk 6 above)
Comment (gargipanatula) [reply internal/credsretriever/refreshing_cache.go:293]: Discussed offline - will consider as a post-merge enhancement.

**Review by SaranBalaji90 — APPROVED** [review]

File: internal/validation/jwk_store.go (hunk 3 above)
Comment (gargipanatula) [reply internal/validation/jwk_store.go:33]: Discussed offline - this may be too protective, will simplify this as a post-merge enhancement.
